### PR TITLE
Obtain participant VCs from Identity Hub

### PR DESCRIPTION
**Pull request by algattik**
## What this PR changes/adds

During participant onboarding, registration service retrieves participant verifiable credentials.

## Why it does that

Enable follow-up items that apply policies (https://github.com/agera-edc/MinimumViableDataspace/issues/11)

## Further notes

Applying any policy on obtained verifiable credentials is out of scope of this work item.

Added participant state transitions to reflect the fact that the call might fail.

## Linked Issue(s)

https://github.com/agera-edc/MinimumViableDataspace/issues/12

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java [hunk 1]
```diff
@@ -129,8 +130,8 @@ public ParticipantStore participantStore() {
     }
 
     @Provider(isDefault = true)
-    public CredentialsVerifier credentialsVerifier() {
-        return new DummyCredentialsVerifier();
+    public ParticipantVerifier participantVerifier(ServiceExtensionContext context) {
+        return new DefaultParticipantVerifier(context.getMonitor(), context.getService(DidResolverRegistry.class), context.getService(CredentialsVerifier.class));
     }
```
Comment (bscholtes1A) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java:135]: Are we sure this wont lead to Dependency Resolution issues? 

According to what we described with Paul in the doc here https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/docs/developer/dependency_resolution.md#provide-defaults this does not seem right, as `DidResolverRegistry` and `CredentialsVerifier` might not have been initialized yet when this default provider is called.

Instead I would see something like this:

~~~
@Inject(required = false)
private ParticipantVerifier participantVerifier;

public void initialize(ServiceExtensionContext context) {
   ...
   if(participantVerifier == null) {
      context.registerService(ParticipantVerifier.class, new DefaultParticipantVerifier(...));
   }
}
~~~

**Review by bscholtes1A — CHANGES_REQUESTED** [review]

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java (hunk 1 above)
Comment (algattik) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java:135]: We do have `@Requires({PrivateKeyResolver.class, OkHttpClient.class, DidResolverRegistry.class, CredentialsVerifier.class})`, so we are guaranteed that the collaborators are initialised before this is called.

Besides, how would your suggested fix help if DidResolverRegistry and CredentialsVerifier were not initialised for some unexpected reason?

Comment (bscholtes1A) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java:135]: > We do have `@Requires({PrivateKeyResolver.class, OkHttpClient.class, DidResolverRegistry.class, CredentialsVerifier.class})`, so we are guaranteed that the collaborators are initialised before this is called.
> 

I am pretty sure that this statement is only true for what is located in the `initialize` method. 

For my proposed fix, this would look like this:

~~~
@Inject(required = false)
private DidResolverRegistry didResolverRegistry;

@Inject(required = false)
private CredentialsVerifier credentialsVerifier;

public void initialize(ServiceExtensionContext context) {
   ...
   if(participantVerifier == null) {
      Objects.requireNotNull(didResolverRegistry, "DidResolverRegistry is mandatory for creating DefaultParticipantVerifier");
      Objects.requireNotNull(credentialsVerifier, "CredentialsVerifier is mandatory for creating DefaultParticipantVerifier");
      context.registerService(ParticipantVerifier.class, new DefaultParticipantVerifier(context.getMonitor(), didResolverRegistry, credentialsVerifier));
   }
}
~~~

Comment (paullatzelsperger) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java]: the problem that @bscholtes1A points out is that the default provider methods are typically called during the `inject` phase of the extension lifecycle. Here, it will work, more by chance than anything else, because the corresponding `@Inject` is coming from the same extension, but as soon as that is in a different extension it may not work anymore, because `@Provider` methods's dependencies are not yet considered by the DI mechanism. 

In other words, the `@Requires` and `@Provides` annotations are not considered when invoking a default provider.

For now, the safest way is to move the provider method into its own extension. I'd also remove the `isDefault` annotation - it was never intended to achieve extensibility. That's what extensions are for.

Comment (algattik) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/AuthorityExtension.java]: Separated extension

**Review by bscholtes1A — APPROVED** [review]

**Review by paullatzelsperger — APPROVED** [review]
